src/data/Wfin.lean

```diff
@@ -0,0 +1,199 @@
+def depth : Wfin ar → ℕ
+| ⟨a, f⟩ := finset.sup finset.univ (λ n, depth (f n)) + 1
+
+lemma not_depth_le_zero (t : Wfin ar) : ¬ t.depth ≤ 0 :=
+by { cases t, apply nat.not_succ_le_zero }
+
+lemma depth_lt_depth_mk (a : α) (f : fin (ar a) → Wfin ar) (i : fin (ar a)) :
+  depth (f i) < depth ⟨a, f⟩ :=
+nat.lt_succ_of_le (finset.le_sup (finset.mem_univ i))
+
+end Wfin
+
+/-!
+Show that `Wfin` types are encodable.
+
+We define an auxiliary type `Wfin' ar n` of trees of depth at most `n`, and
+then we show by induction on `n` that these are all encodable. These auxiliary
+constructions are not interesting in and of themselves, so we mark them as
+`private`.
+-/
+
+namespace encodable
+
+@[reducible] private def Wfin' {α : Type*} (ar : α → ℕ) (n : ℕ) :=
+{ t : Wfin ar // t.depth ≤ n}
+
+variables {α : Type*} {ar : α → ℕ}
+
+private def encodable_zero : encodable (Wfin' ar 0) :=
+let f    : Wfin' ar 0 → empty := λ ⟨x, h⟩, absurd h (Wfin.not_depth_le_zero _),
+    finv : empty → Wfin' ar 0 := by { intro x, cases x} in
+have ∀ x, finv (f x) = x, from λ ⟨x, h⟩, absurd h (Wfin.not_depth_le_zero _),
+encodable.of_left_inverse f finv this
+
+private def f (n : ℕ) : Wfin' ar (n + 1) → Σ a : α, fin (ar a) → Wfin' ar n
+| ⟨t, h⟩ :=
+  begin
+    cases t with a f,
+    have h₀ : ∀ i : fin (ar a), Wfin.depth (f i) ≤ n,
+      from λ i, nat.le_of_lt_succ (lt_of_lt_of_le (Wfin.depth_lt_depth_mk a f i) h),
+    exact ⟨a, λ i : fin (ar a), ⟨f i, h₀ i⟩⟩
+  end
+
+private def finv (n : ℕ) :
+  (Σ a : α, fin (ar a) → Wfin' ar n) → Wfin' ar (n + 1)
+| ⟨a, f⟩ :=
+  let f' := λ i : fin (ar a), (f i).val in
+  have Wfin.depth ⟨a, f'⟩ ≤ n + 1,
+    from add_le_add_right (finset.sup_le (λ b h, (f b).2)) 1,
+  ⟨⟨a, f'⟩, this⟩
+
+variable [encodable α]
+
+private def encodable_succ (n : nat) (h : encodable (Wfin' ar n)) :
+  encodable (Wfin' ar (n + 1)) :=
+encodable.of_left_inverse (f n) (finv n)
+  (by { intro t, cases t with t h, cases t with a fn, reflexivity })
+
+/-- `Wfin` is encodable. -/
+instance : encodable (Wfin ar) :=
+begin
+  haveI h' : Π n, encodable (Wfin' ar n) :=
+    λ n, nat.rec_on n encodable_zero encodable_succ,
+  let f    : Wfin ar → Σ n, Wfin' ar n   := λ t, ⟨t.depth, ⟨t, le_refl _⟩⟩,
+  let finv : (Σ n, Wfin' ar n) → Wfin ar := λ p, p.2.1,
+  have : ∀ t, finv (f t) = t, from λ t, rfl,
+  exact encodable.of_left_inverse f finv this
+end
+
+end encodable
+
+/-!
+As an example, we show that the type of propositional formulas with variables
+labeled from an encodable type is encodable.
+
+The strategy is to define a type of labels corresponding to the constructors.
+From the definition (using `sum`, `unit`, and an encodable type), Lean can infer
+that it is encodable. We then define a map from propositional formulas to the
+corresponding `Wfin` type, and show that map has a left inverse.
+
+We mark the auxiliary constructions `private`, since their only purpose is to
+show encodability.
+-/
+
+namespace Wfin_example
```

Comment:
Does this belong in `tests` or perhaps `docs`?

Reply:
Yeah, I'm not sure where exactly they belong, but they're not ideal here. At the very least, there shouldn't be any global simp lemmas declared.

Reply:
That said, I do think the example is useful to show how to use the API. So `docs` is probably a good home.